streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryWindowStore.java

```diff
@@ -279,39 +275,74 @@ private static Bytes getKey(final Bytes keyBytes) {
 
     }
 
-    private abstract class InMemoryWindowStoreIteratorWrapper implements Comparable<InMemoryWindowStoreIteratorWrapper> {
+    private WrappedInMemoryWindowStoreIterator registerNewWindowStoreIterator(final Bytes key,
+                                                                              final Iterator<Map.Entry<Long, ConcurrentNavigableMap<Bytes, byte[]>>> segmentIterator) {
+        final Bytes keyFrom = retainDuplicates ? wrapForDups(key, 0) : key;
+        final Bytes keyTo = retainDuplicates ? wrapForDups(key, Integer.MAX_VALUE) : key;
+
+        final WrappedInMemoryWindowStoreIterator iterator =
+            new WrappedInMemoryWindowStoreIterator(keyFrom, keyTo, segmentIterator, openIterators::remove);
+
+        openIterators.add(iterator);
+        return iterator;
+    }
+
+    private WrappedWindowedKeyValueIterator registerNewWindowedKeyValueIterator(Bytes keyFrom,
+                                                                                   Bytes keyTo,
+                                                                                   final Iterator<Map.Entry<Long, ConcurrentNavigableMap<Bytes, byte[]>>> segmentIterator) {
+        if (retainDuplicates && !(keyFrom == null || keyTo == null)) {
+            keyFrom = wrapForDups(keyFrom, 0);
+            keyTo = wrapForDups(keyTo, Integer.MAX_VALUE);
```

Comment:
Ack

Reply:
Turns out I was wrong. Since in java all parameters except parameters of primitive types are references, modifications on the state of the objects the parameters point to are visible outside the method (what I meant with side-effects) also if the parameters are declared `final`. With `final` on parameters one can merely avoid modifications of the references stored in the parameters (not the modifications on the state of the object) within the method. I still think that `final` is great and that putting final on parameters makes code a bit better. Just wanted to correct my statement. Sorry for the confusion.

Reply:
No worries, changed anyways b/c I agree it's a better choice stylistically/for code readability :)